### ISSUE TITLE
Style transfer not much

### ISSUE THREAD

**Issue by deeprun** [jcjohnson/fast-neural-style#62]
I am using this network to train the following to style images. These two images were used in your slow-neural-style and achieved impressive results. The training command was 

th train.lua -h5_file COCO_dataset.h5 -style_image  images/addtional_style_image
s/frida_kahlo.jpg -style_image_size 384  -content_weights 1.0 -style_weights 5.0
 -checkpoint_name checkpoint -gpu 0

Training image set is full COCO train and val. Training completed 40000 iterations. 

![frida_kahlo](https://cloud.githubusercontent.com/assets/23156277/19918016/7952f63e-a085-11e6-94f0-d841a5198925.jpg)
![escher_sphere](https://cloud.githubusercontent.com/assets/23156277/19918020/7dca3e70-a085-11e6-90e5-47d7e99671ba.jpg)

After training, I applied it to the following image:
![original](https://cloud.githubusercontent.com/assets/23156277/19918092/efd50748-a085-11e6-96f0-016f5e72a352.png)

Here are the output
![frida](https://cloud.githubusercontent.com/assets/23156277/19918094/f72805a4-a085-11e6-94bb-9908326ad67c.png)
![sphere](https://cloud.githubusercontent.com/assets/23156277/19918096/fa253eb6-a085-11e6-91ed-cc4e3491b65e.png)

As you can see, there is some color changes, but the output images do not have the same "distortion" effect as in the original neural style. How can I improve it?

**Comment by htoyryla** [comment]
This is an example of what I have written about earlier. The new non-iterative feed-forward style transfer methods (this includes texture_nets as well as fast-neural-style) work well with small scale style features but I have not really been able to produce results which would display the essential large-scale style features of the style model in a recognizable way. More style weight tends to bring about a "colored mesh" which has some similarity to the original style yet scaled down to details. I have tried adjusting style scale and the layers used, but without the effect I was looking for. One can certainly create great styles, but the results are in many cases essentially different from the original. Trying to actually reproduce a style that depends not only on details but large scale stylistic features, too, does not seem to succeed.

I had the same difficulty with texture_nets. I wonder if the original neural-style using iterations to approach an optimum of losses is inherently more flexible. The new one-go solutions are trying to model the same behaviour through training, but does this really work for all style transfer cases. It looks like not. Perhaps there are limits as to how much the image can be changed in one go, as opposed to a series of iterative changes?

**Comment by deeprun** [comment]
I found the models provided by Justin had better quality than the ones I trained myself. Need advice about how to making this working better.

**Comment by deeprun** [comment]
@htoyryla I could not find your previous post...

**Comment by htoyryla** [comment]
Never mind... I have mentioned this thought somewhere but there was nothing more than I wrote above. And it is just a question: is the one-go style transfer inherently more limited than an iterative process or not. 

Looking for ways to create great models is completely another thing. It requires experimentation and it takes time because of the training. I guess fast-neural-style is hard to master because there are so many options, and it is very slow to get familiar with them. Fast to use, slow to learn.

Some more thoughts. With texture_nets, it looked to me like some style images made great models without any tweaking, and some styles could not be reproduced at all. The author, Dmitry Ulyanov seemed to have a similar experience. So perhaps here too, if one wants to create great styles, one should look for style images that make great styles (even if the style might not look so much like the original) rather than try to create styles that are difficult (or maybe impossible) to create.

On the other hand, for me personally this question of inherent limits is more interesting than creating great styles.

**Comment by htoyryla** [comment]
Here's an example. This work by Matisse usually works well as a style. It did in texture_nets and it does in fast-neural-style (although I liked the texture_nets result better... and know that quite likely the fast-neural-style results can be improved).

These images also demonstrate how the style features are scaled down to miniature scale in the result. Texture_nets did exactly the same, and there was no easy way to change it. Probably this miniaturization of style features is good in the sense that the model then works better with all kinds of material (easier to look like the content image).

![cy](https://cloud.githubusercontent.com/assets/15064373/19966386/33a0910a-a1d4-11e6-8aa4-d6bfeb73ed43.jpg)

![cy256-svanstein-1z](https://cloud.githubusercontent.com/assets/15064373/19966416/614921d0-a1d4-11e6-9ce2-9e868626c80a.png)

**Comment by htoyryla** [comment]
For comparison, I trained again using style_image_size 960, here's the result. It is different, but not in the sense that the style scale would have changed in a clearly understandable way.

![cy256-svanstein-sis-1z](https://cloud.githubusercontent.com/assets/15064373/19977287/c03f30cc-a1fb-11e6-9852-678751d1565e.png)

**Comment by deeprun** [comment]
Nice. Wereboth images trained by fast-neural-style?

**Comment by htoyryla** [comment]
Yes. What I wanted to show that when the style image is really suitable, you get a reasonably well working style training with more or less the defaults. In the second image I used style_image_size 960, and now I have trained yet another using style_image_size 160. This demonstrates the effect of style_image_size, how it affects the distribution of style features but not so much their scale.

![cy256-svanstein-sis-2z](https://cloud.githubusercontent.com/assets/15064373/19999210/a81ce06a-a279-11e6-80aa-aafa38fffca9.png)

**Comment by deeprun** [comment]
What does exactly this parameter do? 160 means first scale the image down to 160x160 and then fit to VGG network to compute the gama matrix?

**Comment by deeprun** [comment]
Also, did you train style_weights? If you give a super big weights to style and small weight to content, is the image going to be distorted more?

**Comment by htoyryla** [comment]
Increasing the style weight makes the style features dominate, while the content features become less clear. Like this...  style image size 256 like in my first image above but using a higher style weight (16 instead of 6, content weight 1, image generated after 10000 iterations to save time).

So, not really distorting the shape of content like in the Escher image. I have been thinking that maybe that could be achieved by careful selection of layers. Highest layers should be more flexible as to spatial accuracy. So far, however, my initial tests were not successful. Perhaps if one would also finetune weights per layer.

![cy256-svanstein-sw16-1b](https://cloud.githubusercontent.com/assets/15064373/20028275/8c83b86c-a335-11e6-9178-3f8029988baf.png)

**Comment by ycjing** [comment]
A recent work about stroke size control can be found in http://yongchengjing.com/StrokeControllable

**Comment by ycjing** [comment]
Hi, @htoyryla 
Recently, we propose an approach to exploit one single model to achieve continuous stroke size control after training. With our algorithm, a single trained model can achieve different stroke sizes instead of training separate models for each style image scale.
Your work and discussions have helped us a lot. We really appreciate it. Thanks!  ^_^
We have realized the paper and also the corresponding code at:
https://arxiv.org/abs/1802.07101
https://github.com/LouieYang/stroke-controllable-fast-style-transfer

Welcome to pull requests! Thanks.
